File: A18Array11/A2MaxSumSubArray_PrefixSum.java
package A18Array11;

public class A2MaxSumSubArray_PrefixSum {
    public static int SubOfSubArray(int arr[]) {
        int maxSum = 0;
        int currSum = 0;
        int prefix[] = new int[arr.length];

        prefix[0] = arr[0];

        for (int i = 1; i < prefix.length; i++) {
            prefix[i] = prefix[i - 1] + arr[i];
        }

        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr.length; j++) {
                currSum = i == 0 ? prefix[j] : prefix[j] - prefix[i - 1];
                // System.out.println("i=" + i + " " + "j=" + j + " " + currSum);
            }
            if (maxSum < currSum) {
                maxSum = currSum;
            }

        }

        return maxSum;
    }

    public static void main(String[] args) {
        int arr[] = { 2, 3, 4, 6, 8, 9 };
        System.out.println("maxmum sum : " + SubOfSubArray(arr));
    }
}

File: A18Array11/A2MaxSumSubArray_PrefixSum.js
let arr = [2, 3, 4, 6, 8, 9];
function MaxSumPrefixMethod(arr) {
  let prefixArray = [];
  let maxSum = 0;
  let currSum = 0;
  prefixArray[0] = arr[0];

  for (let i = 1; i < arr.length; i++) {
    prefixArray[i] = prefixArray[i - 1] + arr[i];
  }
  console.log("prefixArray", prefixArray);

  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length; j++) {
      currSum = i == 0 ? prefixArray[j] : prefixArray[j] - prefixArray[i - 1];
    }
    if (maxSum < currSum) {
      maxSum = currSum;
    }
  }
  console.log(maxSum);
}

MaxSumPrefixMethod(arr);
